ctowasm/src/modules/binary_tree/index.ts

```diff
@@ -0,0 +1,184 @@
+import {ModulesGlobalConfig, SharedWasmGlobalVariables} from "~src/modules";
+import {NULL_PTR_ADR, voidDataType} from "~src/modules/constants";
+import {Module, ModuleFunction} from "~src/modules/types";
+import {StructDataType} from "~src/parser/c-ast/dataTypes";
+import {
+    convertFloatToCStyleString,
+    extractCStyleStringFromMemory, getAddressOfRegisteredObj,
+    getExternalFunction, loadObjectFromRegistry,
+    storeObjectInMemoryAndRegistry
+} from "~src/modules/util";
+import {MemoryBlock} from "~src/modules/source_stdlib/memory";
+
+// the name that this module is imported into wasm by,
+// as well as the include name to use in C program file.
+export const binaryTreeLibraryModuleImportName = "binary_tree";
+
+export class BinaryTreeLibrary extends Module {
+    moduleDeclaredStructs: StructDataType[];
+    moduleFunctions: Record<string, ModuleFunction>;
+    sharedWasmGlobalVariables: SharedWasmGlobalVariables;
+
+    constructor(
+        memory: WebAssembly.Memory,
+        functionTable: WebAssembly.Table,
+        allocatedBlocks: Map<number, number>,
+        freeList: MemoryBlock[],
+        objectReferenceRegistry: Map<number, any>,
+        config: ModulesGlobalConfig,
+        sharedWasmGlobalVariables: SharedWasmGlobalVariables,
+    ) {
+        super(memory, functionTable, allocatedBlocks, freeList, objectReferenceRegistry, config, sharedWasmGlobalVariables);
+        this.sharedWasmGlobalVariables = sharedWasmGlobalVariables;
+        this.moduleDeclaredStructs = [];
+        this.moduleFunctions = {
+            make_tree: {
+                parentImportedObject: binaryTreeLibraryModuleImportName,
+                functionType: {
+                    type: "function",
+                    parameters: [
+                        {
+                            type: "primary",
+                            primaryDataType: "signed int"
```

Comment:
do specify somewhere in the docs that this defers from normal `make_tree` source academy module as you are only able to take in signed ints as your node value, whereas the module itself could have any JS type